File: cpp/medium/0015_3sum.cpp
// https://leetcode.com/problems/3sum/description/
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> triplets;
        int l, r;
        for(int iter = 0; iter < nums.size(); iter++){
            if(iter > 0 && nums[iter] == nums[iter - 1]){
                // cout << "dupe" << endl;
            } else{
                l = iter + 1;
                r = nums.size() - 1;
                while(l < r){
                    if(nums[iter] + nums[l] + nums[r] == 0){
                        triplets.push_back({nums[iter], nums[l], nums[r]});
                        l++;
                        while(l < r && nums[l] == nums[l - 1]){
                            l++;
                        }
                    } else if(nums[iter] + nums[l] + nums[r] > 0){
                        r--;
                    } else{
                        l++;
                    }
                }
            }
        }
        return triplets;
    }
};

File: python/medium/0015_3sum.py
# https://leetcode.com/problems/3sum/description/
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        prev = nums[0]
        for i in range(0, len(nums)):
            x = nums[i]
            if prev != x or i == 0:
                left = i + 1
                right = len(nums) - 1
                while left < right:
                    y = nums[left]
                    z = nums[right]
                    if x + y + z == 0:
                        res.append([x, y, z])
                        while y == nums[left] and left < right:
                            left += 1
                    elif x + y + z < 0:
                        left += 1
                    else:
                        right -= 1
            prev = x
        return res
